blog/2024-03-15-reflex-architecture.md

```diff
@@ -0,0 +1,166 @@
+
+On the one hand, there are frameworks like [Django](https://www.djangoproject.com/) and [Flask](https://flask.palletsprojects.com/) that are great for building web apps, but they only handle the backend. You still need to learn JavaScript and a frontend framework, as well as writing a lot of boilerplate code to connect the frontend and backend. 
+
+On the other hand, there are pure Python libraries like [Dash](https://dash.plotly.com/) and [Streamlit](https://streamlit.io/) that are great for building data visualization apps. These can be great for small projects, but they are limited to a specific use case and don't have the features and performance to build a full web app.
+
+We wanted to build to bridge this gap by creating a framework that is easy and intuitive, while remaining flexible and powerful to support any app.
+
+## The Reflex Architecture
+
+Reflex apps compile down to a [Next.js](https://github.com/vercel/next.js) frontend app and a [FastAPI](https://github.com/tiangolo/fastapi) backend app. Only the UI is compiled to Javascript; all the app logic and state management stays in Python and is run on the server. Reflex uses websockets to send events from the frontend to the backend, and to send state updates from the backend to the frontend.
+
+When building Reflex, we wanted the final product to look and feel like a traditional web app to the end user, while still being easy to build and maintain for the developer. We chose to leverage the existing web ecosy
+
+The diagram below shows a detailed overview of how a Reflex app works.
+We'll go through each part in more detail in the following sections.
+
+![Reflex Architecture](/architecture.png)
+
+```python eval
+rx.box(height="1em")
+```
+
+### Frontend (Components)
+
+Reflex frontends are built using components that can be composed together to create complex UIs.
+
+```python
+def index():
+    return rx.hstack(
+        rx.avatar(src=f"https://github.com/{GithubState.username}.png"),
+        rx.input(
+            placeholder="Enter your username",
+            on_blur=GithubState.set_username,
+        ),
+    )
+```
+
+In our example app, we have components such as `rx.hstack`, `rx.avatar`, and `rx.input`. Under the hood, these components compile down to React components.
+
+For example, the above code compiles down to the following React code:
+
+```jsx
+<HStack>
+    <Avatar src="https://github.com/{GithubState.username}.png">
+    <Input
+        placeholder="Enter your username"
+        onBlur={GithubState.set_username}
+    >
+</HStack>
+```
+
+We chose React because it is a popular library with a huge ecosystem. Our goal isn't to recreate the web ecosystem, but to make it accessible to Python developers.
+
+This also let's our users bring their own components if we don't have a component they need.
+
+### Backend (State)
+
+In Reflex only the frontend compiles to Javascript, while all the state and logic stays in Python and is run on the server.
```

Comment:
It's great to say the backend runs on the server, but we did not mention where the frontend code runs in the section above

Reply:
added a note on this